### ISSUE TITLE
yolo + stanford cars dataset

### ISSUE THREAD

**Issue by x0rzkov** [AlexeyAB/darknet#5352]
Hi guys,

Hope you are all well !

Main goal:
- recognise a car by its shape as a whole 
- exclude pictures from the trunk, engine, parts of the car, opened doors

What has been done:
- [x] Downloaded the stanford-cars dataset (ref [link](https://ai.stanford.edu/~jkrause/cars/car_dataset.html))
- [x] Converted annotation to yolo format
- [x] train with darknet 7000 images of whole cars 
- [x] Created a colab with the model generated (ref [link](https://drive.google.com/file/d/1iDtkBniWRJthOISARbMBMlR51eSCrFtm/view?usp=sharing))

Current problem:
It detects has car a picture like this one:
![download](https://user-images.githubusercontent.com/56916043/80344155-683c1b00-8867-11ea-8b35-e483c0a65369.png)

Question:
- As any picture from the dataset annotate a light as a car, what should I tweak in the config to get only the whole car detected ?

Thanks in advance for any inputs or insights on that topic.

Cheers,
X

**Comment by x0rzkov** [comment]
[obj.names](http://35.179.44.166:8888/yolo-cars-opencv/obj.names)
[yolo-cars-opencv.cfg](http://35.179.44.166:8888/yolo-cars-opencv/yolo-cars-opencv.cfg)
[yolo-cars-opencv_last.weights](http://35.179.44.166:8888/yolo-cars-opencv/yolo-cars-opencv_last.weights)

**Comment by x0rzkov** [comment]
@AlexeyAB any idea what could be wrong with my model ?

**Comment by LukeAI** [comment]
If your entire training set is whole pictures of cars with no counter-examples then your network will learn that all pictures are of cars. your training data needs to be reflection of the test data. you could try using bdd100K or some other large datasets.

**Comment by x0rzkov** [comment]
So do you mean that I should add 1 or 2 classes like truck or motorbikes ?

**Comment by LukeAI** [comment]
not necessarily. specifically, your training data should reflect your testing data. If you are going to have trucks and motorbikes in your test data, then yes. If you are going to have cars that don't occupy the entire frame at test time, then you should have them in test data as well. if you train with only images of cars that occupy the entire frame, then you will only be able to detect cars that occupy the entire frame.

**Comment by AlexeyAB** [comment]
https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> for each object which you want to detect - there must be at least 1 similar object in the Training dataset with about the same: shape, side of object, **relative size**, angle of rotation, tilt, illumination. So desirable that your training dataset include images with objects at diffrent: scales, rotations, lightings, from different sides, on different backgrounds - you should preferably have 2000 different images for each class or more, and you should train 2000*classes iterations or more

**Comment by x0rzkov** [comment]
@AlexeyAB pictures are not cropped as I provide bounding boxes. SOrry, I am confused. What should I do ? Add random pictures ?

**Comment by AlexeyAB** [comment]
Provide as much as possible
* cfg-file that you used
* training command
* chart.png with Loss
* mAP for Training and Validation datasets

**Comment by x0rzkov** [comment]
- [obj.names](http://35.179.44.166:8888/yolo-cars-opencv/obj.names)
- [yolo-obj.cfg](http://35.179.44.166:8888/yolo-obj.cfg)
- [data.tar.gz](http://35.179.44.166:8888/data.tar.gz)
- [cars.data](http://35.179.44.166:8888/cars.data)
- Training command ```darknet detector train data/obj.data yolo-obj.cfg yolov4.conv.137```

I do not have the 2 last one 
- chart.png with Loss
- mAP for Training and Validation datasets

**Comment by AlexeyAB** [comment]
Set
```
train  = data/train.txt
valid  = data/train.txt
```

and run 
`./darknet detector map data/obj.data yolo-obj.cfg backup/yolo-obj_final.weights`

And show output.

----

> As any picture from the dataset annotate a light as a car, what should I tweak in the config to get only the whole car detected ?

https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> desirable that your training dataset include **images with non-labeled objects that you do not want to detect** - negative samples without bounded box (empty .txt files) - use as many images of negative samples as there are images with objects

**Comment by x0rzkov** [comment]
map info:

```bash
 calculation mAP (mean average precision)...
5704
 detections_count = 15407, unique_truth_count = 5702
class_id = 0, name = car, ap = 88.81%   	 (TP = 4928, FP = 666)

 for conf_thresh = 0.25, precision = 0.88, recall = 0.86, F1-score = 0.87
 for conf_thresh = 0.25, TP = 4928, FP = 666, FN = 774, average IoU = 78.59 %

 IoU threshold = 50 %, used Area-Under-Curve for each unique Recall
 mean average precision (mAP@0.50) = 0.888090, or 88.81 %
Total Detection Time: 219 Seconds

Set -points flag:
 `-points 101` for MS COCO
 `-points 11` for PascalVOC 2007 (uncomment `difficult` in voc.data)
 `-points 0` (AUC) for ImageNet, PascalVOC 2010-2012, your custom dataset
```

I ll add negative images now

**Comment by x0rzkov** [comment]
I have relaunched the training with the negative pictures.
Keep u updated. It says 8 hours to go.

**Comment by x0rzkov** [comment]
No luck, it still recognises a motorbike as a car.

download tarball: [link](http://35.179.44.166:8888/yolo-luc.tar.gz)
